protocols/kad/src/handler.rs

```diff
@@ -780,12 +768,32 @@ where
             return Poll::Ready(event);
         }
 
-        let num_in_progress_outbound_substreams =
-            self.outbound_substreams.len() + self.num_requested_outbound_streams;
-        if num_in_progress_outbound_substreams < MAX_NUM_SUBSTREAMS {
-            if let Some(protocol) = self.requested_streams.pop_front() {
+        'outer: while self.outbound_substreams.len() + self.num_requested_outbound_streams
+            < MAX_NUM_SUBSTREAMS
+        {
+            if let Some((msg, user_data)) = self.pending_substream_requests.pop_front() {
+                // Search for outbound substream waiting to be reused first.
+                for outbound_substream in self.outbound_substreams.iter_mut() {
+                    match std::mem::replace(outbound_substream, OutboundSubstreamState::Poisoned) {
+                        OutboundSubstreamState::Idle(substream, waker) => {
+                            *outbound_substream =
+                                OutboundSubstreamState::PendingSend(substream, msg, user_data);
+                            waker.wake();
+                            continue 'outer;
+                        }
+                        other => {
+                            *outbound_substream = other;
+                        }
+                    }
+                }
```

Comment:
For consistency, can we extract this into a fn like `try_answer_request`?

That should also allow us to get rid of the loop markers.